src/bsoncxx/types/value.hpp

```diff
@@ -154,8 +155,9 @@ namespace types {
 
         ///
         /// @return The type of the underlying BSON value stored in this object.
+        /// If the type is invalid, a disengaged optional is returned.
         ///
-        bsoncxx::type type() const;
+        stdx::optional<bsoncxx::type> type() const;
```

Comment:
What would you think about adding an `BSONCXX_INLINE explicit operator bool() const` to this class that just delegated to calling type and checking the state of the optional. Then you could check whether a value was engaged or not by calling `if (value)`, rather than `if (value.type())`